job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.job.task;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * ThreadPoolExecutor that can be temporarily paused to prevent any threads from
+ * executing new tasks.
+ */
+public class PausableThreadPoolExecutor extends ThreadPoolExecutor {
+
+  private boolean mIsPaused;
+  private int mNumPaused;
+  private ReentrantLock mPauseLock;
+  private Condition mUnpaused;
+
+  /**
+    * Copy of one of the constructors in {@link ThreadPoolExecutor}.
+    *
+    * @param corePoolSize the core pool size
+    * @param maximumPoolSize the maximum pool size
+    * @param keepAliveTime the keep alive time
+    * @param unit the unit
+    * @param workQueue the work queue
+    * @param threadFactory the thread factory
+    */
+  public PausableThreadPoolExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime,
+      TimeUnit unit, BlockingQueue<Runnable> workQueue, ThreadFactory threadFactory) {
+    super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue, threadFactory);
+    mIsPaused = false;
+    mNumPaused = 0;
+    mPauseLock =  new ReentrantLock();
+    mUnpaused = mPauseLock.newCondition();
+  }
+
+  /**
+   * @return number of active threads subtracted by the number of tasks paused
+   */
+  public int getNumActiveTasks() {
+    return super.getActiveCount() - mNumPaused;
```

Comment:
Is it correct that `mPauseLock` is not protecting `mNumPaused`?

Reply:
Access to mNumPaused is not protected but that's okay. Because protect it or not, the value of mNumPaused can be inaccurate. But being slightly inaccurate here is okay.

Reply:
Ok. Does it have to be `volatile` then? I'm not entirely sure how "stale" a non-volatile variable might be.